### PR TITLE
fix(controller): End DAG execution on deadline exceeded error. Fixes #3905

### LINKED ISSUE #3905: DAGs may not abort on `ErrDeadlineExceeded`
## Summary 

When the deadline is exceeded, the DAG should abort and return. Instead, ""Deadline exceeded" is printed as many times as the number of items.

dag.go:

image

## Diagnostics

master

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

 Fixes #3905

### CODE REVIEW

File: workflow/controller/dag.go [hunk 1]
```diff
@@ -432,7 +432,17 @@ func (woc *wfOperationCtx) executeDAGTask(dagCtx *dagContext, taskName string) {
 		}
 
 		// Finally execute the template
-		_, _ = woc.executeTemplate(taskNodeName, &t, dagCtx.tmplCtx, t.Arguments, &executeTemplateOpts{boundaryID: dagCtx.boundaryID, onExitTemplate: dagCtx.onExitTemplate})
+		_, err = woc.executeTemplate(taskNodeName, &t, dagCtx.tmplCtx, t.Arguments, &executeTemplateOpts{boundaryID: dagCtx.boundaryID, onExitTemplate: dagCtx.onExitTemplate})
+		if err != nil {
```
Comment (alexec) [inline workflow/controller/dag.go:436]: this is effectively a copy-and-paste of [steps.go#233](https://github.com/argoproj/argo/blob/master/workflow/controller/steps.go#L233)

File: workflow/controller/dag.go [hunk 2]
```diff
@@ -432,7 +432,17 @@ func (woc *wfOperationCtx) executeDAGTask(dagCtx *dagContext, taskName string) {
...
+			switch err {
+			case ErrDeadlineExceeded:
+				return
+			case ErrParallelismReached:
+			default:
+				woc.log.WithField("taskName", t.Name).WithError(err).Warn("failed to execute expanded task")
```
Comment (alexec) [inline workflow/controller/dag.go:442]: is this the correct behaviour for a dag? steps.go does a bit extra work.

**Review by simster7 — COMMENTED** [review]
Do we need to add this everywhere where `executeTemplate` is called?

**Comment by alexec** [comment]
> Do we need to add this everywhere where `executeTemplate` is called?

Yes. But this is the only place that needs changing AFAIK.

File: workflow/controller/operator_test.go [hunk 3]
```diff
@@ -4684,8 +4684,8 @@ func TestTemplateTimeoutDuration(t *testing.T) {
 		makePodsPhase(t, apiv1.PodPending, controller.kubeclientset, wf.ObjectMeta.Namespace)
 		woc.operate()
 
-		assert.Equal(t, wfv1.NodeFailed, woc.wf.Status.Phase)
-		assert.Equal(t, wfv1.NodeFailed, woc.wf.Status.Nodes.FindByDisplayName("hello-world-dag").Phase)
+		assert.Equal(t, wfv1.NodeError, woc.wf.Status.Phase)
```
Comment (alexec) [inline workflow/controller/operator_test.go:4687]: I'm not sure what is correct here. This code change is exercised by this test. The steps test marks the node as `Error` but the workflows as `Failed`. I think the steps should also mark as `Error` - or both should be `Failed`.

@sarabala1979 thoughts? @jessesuen

Comment (alexec) [reply workflow/controller/operator_test.go:4687]: Oh - for context - this is the new timeout code @sarabala1979 just merged.

Comment (alexec) [reply workflow/controller/operator_test.go]: I think we need to be clear when something is an error and when something is a failure. I'm a bit unclear on this looking at the code.

Comment (sarabala1979) [reply workflow/controller/operator_test.go]: Testcase is wrong. I will fix it. It should check the failed for the step which timed out

Comment (alexec) [reply workflow/controller/operator_test.go]: I've made a change to your implementation to return a specific `ErrTimeout` that allows us to switch based on the result.

However, I'm still not sure if the node and workflow should be failed or errored.

**Review by sarabala1979 — APPROVED** [review]
LGTM

**Pull request merged**